src/NzbDrone.Core/Organizer/FileNameBuilder.cs

```diff
@@ -332,6 +339,58 @@ private void AddMediaInfoTokens(Dictionary<string, Func<TokenMatch, string>> tok
                 m => MediaInfoFormatter.FormatVideoDynamicRange(movieFile.MediaInfo);
         }
 
+        private void AddCustomFormats(Dictionary<string, Func<TokenMatch, string>> tokenHandlers, Movie movie, MovieFile movieFile, List<CustomFormat> customFormats = null)
+        {
+            if (customFormats == null)
+            {
+                var info = new ParsedMovieInfo
+                {
+                    MovieTitle = movie.Title,
+                    SimpleReleaseTitle = movieFile.GetSceneOrFileName().SimplifyReleaseTitle(),
+                    Quality = movieFile.Quality,
+                    Languages = movieFile.Languages,
+                    ReleaseGroup = movieFile.ReleaseGroup,
+                    Edition = movieFile.Edition,
+                    Year = movie.Year,
+                    ImdbId = movie.ImdbId,
+                    ExtraInfo = new Dictionary<string, object>
+                    {
+                        { "IndexerFlags", movieFile.IndexerFlags },
+                        { "Size", movieFile.Size },
+                        { "Filename", System.IO.Path.GetFileName(movieFile.RelativePath) }
+                    }
+                };
+
+                var allFormats = _formatService.All();
+
+                var matches = new List<CustomFormatMatchResult>();
+
+                foreach (var customFormat in allFormats)
+                {
+                    var tagTypeMatches = customFormat.FormatTags
+                        .GroupBy(t => t.TagType)
+                        .Select(g => new FormatTagMatchesGroup
+                        {
+                            Type = g.Key,
+                            Matches = g.ToDictionary(t => t, t => t.DoesItMatch(info))
+                        })
+                        .ToList();
+
+                    matches.Add(new CustomFormatMatchResult
+                    {
+                        CustomFormat = customFormat,
+                        GroupMatches = tagTypeMatches
+                    });
+                }
+
+                customFormats = matches.Where(m => m.GoodMatch)
+                                       .Select(r => r.CustomFormat)
+                                       .ToList();
```

Comment:
Recreated the calculation for movieFile formats here as bringing in the calculation service creates IOC issues. This is because the calculation service uses MovieService and ParsingService for history and blacklist parsing which loop back to using FileNameBuilder creating a circular reference. 

Open to any other solution.

Reply:
What about if we factor out the CF calculation service into a parsed movie info mapping service and the calculation service itself, which only accepts a parsed movie info. Then we lose the circular dependency I think, but you would have to manually call the mapping service then the calculation service

Reply:
Yea that could work if we can live with the extra calls